### PR TITLE
[kvexec] customized operator for count aggregation

### PR DESCRIPTION
**Pull request by max-hoffman**
Operators that count the number of rows in a relation don't have to deserialize those KV's from storage (ex: `select count(y) from xy where x > 1`). There are some circumstances where we have to check for field nullability, but otherwise we can just count the KV's returned by the source iterator.

### CODE REVIEW

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>17c017f</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>17c017f</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_p95 | to_latency_p95 | is_faster |
|---------------------|------------------|----------------|-----------|
| tpcc-scale-factor-1 | 59.99            | 62.19          | 0         |


|      test_name      | server_name |              server_version              |  tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|-------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | b2f0662cb83c1521b5fadf03e2be0d2bcd14a9d5 | 39.37 | tpcc-scale-factor-1 | dolt        | 17c017f9ee90a5ccff25a104a89b40fb2641eaa4 | 39.75 | 0         |

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>f8fc91f</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>f8fc91f</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 2.97                | 1.18              | 1         |
| groupby_scan         | 17.32               | 17.01             | 0         |
| index_join           | 2.66                | 2.66              | 0         |
| index_join_scan      | 2.14                | 2.14              | 0         |
| index_scan           | 54.83               | 54.83             | 0         |
| oltp_point_select    | 0.3                 | 0.29              | 0         |
| oltp_read_only       | 5.88                | 5.88              | 0         |
| select_random_points | 0.65                | 0.65              | 0         |
| select_random_ranges | 0.83                | 0.8               | 0         |
| table_scan           | 54.83               | 54.83             | 0         |
| types_table_scan     | 142.39              | 144.97            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 5.88                | 5.88              | 0         |
| oltp_insert           | 2.97                | 2.97              | 0         |
| oltp_read_write       | 12.08               | 11.87             | 0         |
| oltp_update_index     | 3.02                | 2.97              | 0         |
| oltp_update_non_index | 2.97                | 2.91              | 0         |
| oltp_write_only       | 6.09                | 6.09              | 0         |
| types_delete_insert   | 6.43                | 6.43              | 0         |

**Comment by max-hoffman** [comment]
#benchmark

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      test_name      | from_latency_p95 | to_latency_p95 | is_faster |
|---------------------|------------------|----------------|-----------|
| tpcc-scale-factor-1 | 61.08            | 61.08          | 0         |


|      test_name      | server_name |              server_version              | tps  |      test_name      | server_name |              server_version              |  tps  | is_faster |
|---------------------|-------------|------------------------------------------|------|---------------------|-------------|------------------------------------------|-------|-----------|
| tpcc-scale-factor-1 | dolt        | a50b1b71fe247d8c751d9ba2e8914c743f9e09e2 | 39.0 | tpcc-scale-factor-1 | dolt        | bb31dbd9e71e1a51b4e69f1cf04bed22211d8fbd | 39.84 | 0         |

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 |      read_tests      | from_latency_median | to_latency_median | is_faster |
|----------------------|---------------------|-------------------|-----------|
| covering_index_scan  | 2.91                | 1.21              | 1         |
| groupby_scan         | 17.01               | 17.01             | 0         |
| index_join           | 2.66                | 2.61              | 0         |
| index_join_scan      | 2.14                | 2.18              | 0         |
| index_scan           | 54.83               | 54.83             | 0         |
| oltp_point_select    | 0.3                 | 0.3               | 0         |
| oltp_read_only       | 5.99                | 5.99              | 0         |
| select_random_points | 0.65                | 0.65              | 0         |
| select_random_ranges | 0.83                | 0.8               | 0         |
| table_scan           | 55.82               | 54.83             | 0         |
| types_table_scan     | 142.39              | 142.39            | 0         |


|      write_tests      | from_latency_median | to_latency_median | is_faster |
|-----------------------|---------------------|-------------------|-----------|
| oltp_delete_insert    | 5.99                | 5.99              | 0         |
| oltp_insert           | 3.02                | 2.97              | 0         |
| oltp_read_write       | 12.08               | 12.08             | 0         |
| oltp_update_index     | 3.02                | 3.02              | 0         |
| oltp_update_non_index | 2.97                | 2.97              | 0         |
| oltp_write_only       | 6.21                | 6.21              | 0         |
| types_delete_insert   | 6.55                | 6.43              | 0         |

File: go/libraries/doltcore/sqle/kvexec/count_agg.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+func (l *countAggKvIter) Next(ctx *sql.Context) (sql.Row, error) {
+	// will return one value
+	if l.done {
+		return nil, io.EOF
+	}
+	var cnt int64
+	for {
+		k, v, err := l.srcIter.Next(ctx)
+		if err == io.EOF {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/kvexec/count_agg.go:82]: Not handling any other error here

File: go/libraries/doltcore/sqle/kvexec/count_agg_test.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		{
+			name: "accept in subquery with non-nil scope",
+			setup: []string{
+				"create table xy (x int primary key, y int, key y_idx(y))",
+			},
+			query:     "select  1, (select count(y) from xy) cnt",
+			doRowexec: true,
+		},
+		{
+			name: "reject complex count expressi	on",
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/kvexec/count_agg_test.go:83]: spacing

**Review by zachmu — APPROVED** [review]
LGTM

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>ff89161</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>ff89161</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>4fc3498</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>4fc3498</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d35a051</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d35a051</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>100.000000 to 100.000000</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>20c6e26</TD>
<TD>ok</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>20c6e26</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>100.0</TD>
</TR>
</table>

**Pull request merged**